src/gui/Icons.cpp

```diff
@@ -211,3 +217,147 @@ Icons* Icons::instance()
+        return databaseIcons()->icon(entry->iconNumber()).toImage();
+    } else {
+        Q_ASSERT(entry->database());
+
+        if (entry->database()) {
+            return Icons::customIcon(entry->database(), entry->iconUuid());
+        } else {
+            return QImage();
+        }
+    }
+}
+
+QPixmap Icons::entryIconPixmap(const Entry* entry, IconSize size)
+{
+    QPixmap icon(size, size);
+    if (entry->iconUuid().isNull()) {
+        icon = databaseIcons()->icon(entry->iconNumber(), size);
+    } else {
+        Q_ASSERT(entry->database());
+        if (entry->database()) {
+            icon = Icons::customIconPixmap(entry->database(), entry->iconUuid(), size);
+        }
+    }
+
+    if (entry->isExpired()) {
+        icon = databaseIcons()->applyBadge(icon, DatabaseIcons::Badges::Expired);
+    }
+
+    return icon;
+}
+
+QImage Icons::groupIcon(const Group* group)
+{
+    if (group->iconUuid().isNull()) {
+        return databaseIcons()->icon(group->iconNumber()).toImage();
+    } else {
+        Q_ASSERT(group->database());
+        if (group->database()) {
+            return Icons::customIcon(group->database(), group->iconUuid());
+        } else {
+            return QImage();
+        }
+    }
+}
+
+QPixmap Icons::groupIconPixmap(const Group* group, IconSize size)
+{
+    QPixmap icon(size, size);
+    if (group->iconUuid().isNull()) {
+        icon = databaseIcons()->icon(group->iconNumber(), size);
+    } else {
+        Q_ASSERT(group->database());
+        if (group->database()) {
+            icon = Icons::customIconPixmap(group->database(), group->iconUuid(), size);
+        }
+    }
+
+    if (group->isExpired()) {
+        icon = databaseIcons()->applyBadge(icon, DatabaseIcons::Badges::Expired);
+    }
+#ifdef WITH_XC_KEESHARE
+    else if (KeeShare::isShared(group)) {
+        icon = KeeShare::indicatorBadge(group, icon);
+    }
+#endif
+
+    return icon;
+}
+
+QString Icons::imageReaderFilter()
+{
+    const QList<QByteArray> formats = QImageReader::supportedImageFormats();
+    QStringList formatsStringList;
+
+    for (const QByteArray& format : formats) {
+        for (char codePoint : format) {
+            if (!QChar(codePoint).isLetterOrNumber()) {
+                continue;
+            }
+        }
+
+        formatsStringList.append("*." + QString::fromLatin1(format).toLower());
+    }
+
+    return formatsStringList.join(" ");
+}
+
+QByteArray Icons::getBytes(const QImage& image)
+{
+#if QT_VERSION >= QT_VERSION_CHECK(5, 10, 0)
+    return QByteArray(reinterpret_cast<const char*>(image.bits()), static_cast<int>(image.sizeInBytes()));
+#else
+    return QByteArray(reinterpret_cast<const char*>(image.bits()), image.byteCount());
+#endif
+}
+
+QByteArray Icons::saveToBytes(const QImage& image)
+{
+    QByteArray ba;
+    QBuffer buffer(&ba);
+    buffer.open(QIODevice::WriteOnly);
+    // TODO: check !icon.save()
+    image.save(&buffer, "PNG");
+    buffer.close();
+    return ba;
```

Comment:
@droidmonkey there was 2 methods used for converting a `QImage` in the codebase. The first (`getBytes`) was used when hashing the images in `Metadata`. The second `saveToBytes` is used for save the images. I feel like we might use `saveToBytes` for both, but still not 100% sure about that.

Reply:
I agree with that

Reply:
I removed `getBytes`, we only use `saveToBytes` now.